rfcs/2022-11-04-automatic-component-validation.md

```diff
@@ -0,0 +1,611 @@
+the system before and after the code runs, calculating the validation results at the end.
+
+We do not have an existing mechanism for asserting the _correctness_ of internal telemetry, only
+that specific bits of internal telemetry are being emitted as intended. This is not necessarily a
+technical limitation, but a practical limitation of not yet having gotten to the point of adding
+support to validate the correctness.
+
+What both of these things do not provide, that this proposal does, is a mechanism to _ensure_ that a
+component that can be used has been actually tested against the intended criteria/specifications/etc
+and met them. While this doesn't happen as a result of the compiler/type system, and so is still
+somewhat fallible at the edges -- we have to actually do work to integrate this proposal with CI,
+etc -- it achieves uniform coverage across all components and provides the scaffolding to avoid the
+human fallibility aspect of "oops, we're not actually validating this component".
+
+It also heavily centralizes this validation logic such that any future work around updating what
+gets validated, or exposing more information during the validation step, and so on, can happen as a
+set of changes to a relatively small set of files, rather than, again, a change to every component
+involved, which is exactly where human fallibility has struck us in the past.
+
+## Alternatives
+
+One alternative approach could be to actually build Vector and run an entire configuration, much
+like an integration test, and we would send known data to Vector, collect the output, so on and so
+forth, much like this proposal... but we would simply collect it externally, much like `lading` does
+in our soak tests. Said another way, replace the part of `Runner` that builds a `RunningTopology`
+from a `ConfigBuilder`... and simply generate a real Vector configuration and instead spawn a Vector process.
+
+This approach does solve the issue of isolation of telemetry data rather concretely, as each test
+case would be a new Vector process, with no risk whatsoever of contaminated telemetry events between
+test cases. It also potentially opens up the ability to provide full-fledged configuration files
+where the proposed traits/interfaces are suboptimal and would require complex refactoring to support
+edge cases.
+
+The biggest reason not to go with this approach is that we lose some measure of programmatic control
+over running the Vector topology:
+
+- we would lose the ability to use in-application synchronization primitives for determining when
+  components/tasks were ready (such as whether or not a component was ready, relegating us to trying
+  to, for example, connect to a TCP socket until we finally got a connection)
+- determining/collecting errors could be harder as we would be forced to parse the stdout/stderr of
+  the Vector process in some cases to understand issues with the configuration
```

Comment:
So, admittedly, the reasons to avoid _are_ somewhat weak, because all of the approaches are so similar, IMO, that you're really just trading off one nuance for another.

If I could answer this question in a sort of inverse way of "why is the proposed approach better than the listed alternative?", I would say the following:

**1. Running an isolated Vector process means having to (additionally) build a non-test binary**

This one is the biggest reason not to prefer the alternative. Maybe building the Vector binary from a unit test wouldn't actually be that hard, or maybe it wouldn't be much work to whip up a Makefile target to do the build first and then invoke the test (whether it's a new binary crate, or unit test, or ...). We do that for `make generate-component-docs`, and it's certainly technically possible.

It does strike me, though, that the proposed approach is the one that should yield the fastest turnaround time for revalidating a component during the edit-test loop. We're using code that is already going to end up in the test binary, so we don't have to build/link anything else.

**2. Running everything in-process is more deterministic**

To your point, it's not really an issue with needing special synchronization primitives: we can run the tests and know when state has changed, etc, even just by polling, as you mentioned. Doing so deterministically, in a more event-driven manner, however, would become harder. Without a deterministic approach, such as a normal async task wakeup style of mechanism, we would likely spend extra time per test case waiting to begin sending inputs.

For the sake of argument, let's say we additionally spend somewhere on the order of 20-30ms waiting to determine that the spawned process is ready for us to proceed with the test. At over 100 components, which will likely have _at least_ two test cases (one success, one failure), that's 4-5 extra seconds per test run.

In practice, I would expect that we land somewhere around 1500-2000ms per test case even with my proposed approach, so this additional time is still a small percentage overall... but based on how long we've had to delay some existing unit tests that wait for a topology to start running and listening on a port, my 20-30ms "extra time spent polling" estimate is likely _very_ generous.

Reply:
To answer your questions specifically:

> Would setting up that synchronization, even if it just reduced to polling, be harder than what is described here?

Simply put, no: it wouldn't be harder or more complex to do so in a functionally-equivalent way. It would, as mentioned above, though, simply be less deterministic.

> Would the errors not already need to be parsed by human to determine what went wrong?

In this case, I was thinking of issues that could occur during the translation from code in the unit test figuring out the Vector configuration to build, and then Vector being spawned and trying to load that configuration. Think of all the weird, non-obvious `serde` deserialization issues we've seen. _Serializing_ our configuration types is, arguably, one of our least well-tested aspects of any of the code that we have surrounding Vector's configuration in general. There's no chance of having to contend with any of that if we're simply constructing and using those configuration type values directly.

There's obviously other possible errors that, on balance, would be likely just as easy to grok, regardless of whether we bubble them up in the validation unit test runner or simply dump to console via `panic!` and have to sift through manually. It would be synergistic to make those user-facing errors that Vector emits be as good as a well-tailored error enum, certainly.